### PR TITLE
Update `Deployment.infrastructure` to accept types outside of the core library

### PR DESCRIPTION
**Pull request by madkinsz**
Updates deployment `infrastructure` field type to the `Infrastructure` base class instead of enumerating all types. This is required for the upcoming infrastructure blocks in collections e.g. `ECSTask` and `CloudRunJob`.

#6622 should allow us to do this without any issues.

### Example

```python
from prefect_aws.ecs import ECSTask

Deployment(name="foo", infrastructure=ECSTask(...))
```

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a `fix`, `feature`, `enhancement`, `docs`, or `maintenance` label categorizing the change.

### CODE REVIEW

**Comment by madkinsz** [comment]
@peytonrunyan do you want to give this a test with your work? It'd be nice to include an explicit test for this by defining a dynamic `Infrastructure` type and creating a deployment with it if you're up to it.

**Comment by madkinsz** [comment]
Sorry this should build on #6622, I thought it was merged already but it's not. There's no way this should be passing tests as-is, I'm guessing there's some important coverage missing for deployments.

**Comment by peytonrunyan** [comment]
I'm in!

**Comment by madkinsz** [comment]
@peytonrunyan I rebased this onto #6622 — let me know how it goes. It'd be good to add a unit test here with a fake infrastructure block type if you have the chance.

**Comment by peytonrunyan** [comment]
I was experimenting with a GUI tool to squash commits and it looks like I went back one commit too far. I'm going to hold off on trying to fix this until I've got a slightly better idea of the right solution. Git isn't a strong point for me.

File: src/prefect/orion/database/configurations.py [hunk 1]
```diff
@@ -109,7 +109,12 @@ async def engine(self) -> AsyncEngine:
             if connect_args:
                 kwargs["connect_args"] = connect_args
 
-            engine = create_async_engine(self.connection_url, echo=self.echo, **kwargs)
+            engine = create_async_engine(
+                self.connection_url,
+                echo=self.echo,
+                execution_options={"isolation_level": "AUTOCOMMIT"},
```
Comment (cicdw) [inline src/prefect/orion/database/configurations.py:115]: what are these changes related to? I'm not as familiar with our database configuration but my understanding is that changing the isolation level is a pretty significant change

**Review by cicdw — CHANGES_REQUESTED** [review]
Deployment changes look good to me, but i have a question about the DB change

File: src/prefect/orion/database/configurations.py (hunk 1 above)
Comment (madkinsz) [reply src/prefect/orion/database/configurations.py]: This was a weird merge with `main`. Should be gone now.

File: tests/test_deployments.py [hunk 2]
```diff
@@ -30,6 +30,19 @@ async def test_that_infrastructure_defaults_to_process(self):
...
+        class CustomInfra(Infrastructure):
+            type = "CustomInfra"
+
+            def run(self):
+                return 42
+
+            def preview(self):
+                return "woof!"
+
+        d = Deployment(name="foo", infrastructure=CustomInfra())
```
Comment (madkinsz) [inline tests/test_deployments.py:43]: Can you add a test with `infrastructure=CustomInfra().dict()` too? We need that to ensure it works from YAML.

Comment (madkinsz) [reply tests/test_deployments.py:43]: (or an actual test that loads from a YAML config which would be better)

Comment (madkinsz) [reply tests/test_deployments.py:43]: @peytonrunyan want to make sure you've seen this as we kick off the new week

**Comment by madkinsz** [comment]
Not merged. Something went wrong. Will fix :D

**Comment by madkinsz** [comment]
Alright, I've rebased the history once again.

**Review by anna-geller — APPROVED** [review]
